File: Web/src/domain/team.ts
import type { Player, Team } from './models'
import { getNameAndSurname, getVote } from './player'

/**
 * Funzioni pure su Team, portate da Team.java.
 */

/**
 * Voto totale della squadra.
 *
 * ⚠️ Riproduce fedelmente `Team.addPlayer()`, che arrotonda a un decimale
 * **dopo ogni singolo inserimento**. L'arrotondamento è quindi cumulativo e
 * dipende dall'ordine dei giocatori: sommare tutto e arrotondare alla fine
 * darebbe risultati leggermente diversi da quelli mostrati dall'app Android.
 */
export function getTeamVote(team: Team): number {
  let total = 0
  for (const player of team.players) {
    total = Math.round((total + getVote(player)) * 10) / 10
  }
  return total
}

/** Numero di giocatrici nella squadra: vincolo di bilanciamento del generatore. */
export function countFemales(team: Team): number {
  return team.players.filter((p) => p.gender === 'F').length
}

/** Solo i nomi: "Marco - Luca - Anna". Porta `toStringOnlyName()`. */
export function formatNames(team: Team): string {
  return team.players.map((p) => p.name).join(' - ')
}

/** Nome + iniziale del cognome: "Marco R - Luca B". Porta `toStringShortName()`. */
export function formatShortNames(team: Team): string {
  return team.players
    .map((p) => (p.surname.length > 0 ? `${p.name} ${p.surname[0]}` : p.name))
    .join(' - ')
}

/** Nome + cognome completo. Porta `toStringNameAndSurname()`. */
export function formatFullNames(team: Team): string {
  return team.players
    .map((p) => (p.surname.length > 0 ? `${p.name} ${p.surname}` : p.name))
    .join(' - ')
}

/** Nome + cognome troncato. Porta `toStringNameAndSurnameNChar(int)`. */
export function formatTruncatedNames(team: Team, maxSurnameChars: number): string {
  return team.players.map((p) => getNameAndSurname(p, maxSurnameChars)).join(' - ')
}

/**
 * Numero progressivo della squadra nel torneo (1-based), come mostrato in UI
 * ("TEAM 3"). Porta `Tournament.getNTeamByKey()`, che si basa sull'ordine
 * degli elementi nella lista. Restituisce 0 se la chiave non esiste.
 */
export function getTeamNumber(teams: Team[], teamKey: string): number {
  return teams.findIndex((t) => t.key === teamKey) + 1
}

/** Costruisce una squadra a partire dalle chiavi dei giocatori. */
export function buildTeam(key: string, bracket: string, players: Player[]): Team {
  return { key, bracket, players }
}

File: AppAndroid/app/src/main/java/com/example/myapplication/Team.java
package com.example.myapplication;

import com.example.myapplication.Model.Constants;

import java.util.ArrayList;

public class Team {
    public String key;
    public float totalVote = 0;
    public int points = 0;
    public String bracket = "";
//    float average = 0;
    public ArrayList<Player> players = new ArrayList<>();

    public Team() {}

    public void addPlayer(Player player) {
        this.players.add(player);
        this.totalVote += player.getVote();
        this.totalVote = (float) (Math.round(this.totalVote * 10.0) / 10.0);
//        this.average = (float) (Math.round(this.totalVote / this.players.size() * 10.0) / 10.0);
    }

    public float getTotalVote() {
        return totalVote;
    }

    public void addPlayerByKey(String key) {
        for (Player p : Constants.players) {
            if (p.key.equals(key)) {
                Player temp = new Player();
                temp.setTo(p);
                addPlayer(temp);
            }
        }
    }

    @Override
    public String toString() {
        return  key + " - " + totalVote + " - " + players;
    }

    public String toStringOnlyName() {
        StringBuilder toReturn = new StringBuilder();

        for (int i = 0; i < players.size(); i++) {
            toReturn.append(players.get(i).name);

            if (i != players.size() - 1) {
                toReturn.append(" - ");
            }
        }

        return toReturn.toString();
    }

    public String toStringShortName() {
        StringBuilder toReturn = new StringBuilder();

        for (int i = 0; i < players.size(); i++) {
            toReturn.append(players.get(i).name);

            if (!players.get(i).surname.isEmpty())
                toReturn.append(" ").append(players.get(i).surname.charAt(0));

            if (i != players.size() - 1)
                toReturn.append(" - ");
        }

        return toReturn.toString();
    }

    public String toStringNameAndSurname() {
        StringBuilder toReturn = new StringBuilder();

        for (int i = 0; i < players.size(); i++) {
            toReturn.append(players.get(i).name);

            if (!players.get(i).surname.isEmpty())
                toReturn.append(" ").append(players.get(i).surname);

            if (i != players.size() - 1)
                toReturn.append(" - ");
        }

        return toReturn.toString();
    }

    public String toStringNameAndSurnameNChar(int nChar) {
        StringBuilder toReturn = new StringBuilder();

        for (int i = 0; i < players.size(); i++) {
            toReturn.append(players.get(i).getNameAndSurname(nChar));

            if (i != players.size() - 1)
                toReturn.append(" - ");
        }

        return toReturn.toString();
    }
}
